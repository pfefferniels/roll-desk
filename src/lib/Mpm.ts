import { parse } from "js2xmlparser"

/**
 * A part can be specified as either a given
 * part number or global. This definition is 
 * used in both, MSM and MPM.
 */
export type Part = number | 'global'

type DatedInstruction<T extends string> = {
    readonly type: T
}

/**
 * Maps the <dynamics> element of MPM
 */
export interface Dynamics extends DatedInstruction<'dynamics'> {
    date: number
    volume: number | string
    'transition.to'?: number
}

/**
 * Maps the <tempo> element of MPM
 */
export interface Tempo extends DatedInstruction<'tempo'> {
    'date': number
    'bpm': number
    'beatLength': number
    'transition.to'?: number
    'meanTempoAt'?: number
}

/**
 * Maps the <ornament> element of MPM
 */
export interface Ornament extends DatedInstruction<'ornament'> {
    date: number
    'name.ref': string
    'note.order': string
    'frameLength': number 
    'frame.start': number
    'transition.from'?: number
    'transition.to'?: number
    'scale': number
}

type AnyInstruction = 
    | Tempo 
    | Ornament 
    | Dynamics

export class MPM {
    rawMPM: any

    constructor(nParts = 2) {
        this.rawMPM = {
            "@": {
                xmlns: "http://www.cemfi.de/mpm/ns/1.0"
            },
            performance: {
                "@": {
                    name: '',
                    pulsesPerQuarter: 720
                },
                global: {
                    dated: {
                        'tempoMap': {},
                        'ornamentationMap': {},
                        'imprecisionMap.timing': {}
                    }
                },
                part: Array.from(Array(nParts).keys()).map(i => {
                    return {
                        "@": {
                            name: `part${i}`,
                            number: `${i + 1}`,
                            "midi.channel": `${i}`,
                            "midi.port": "0"
                        },
                        dated: {
                            dynamicsMap: {},
                            asynchronyMap: {},
                            ornamentationMap: {},
                            articulationMap: {}
                        }
                    }
                })
            }
        }
    }

    /**
     * Returns the instructions at a given date.
     * @param date 
     * @param part If not specified, all parts are considered
     */
    instructionsAtDate(date: number, part?: Part) {
        if (part === undefined) {

        }
        else if (part === 'global') {

        }
        else if (typeof part === 'number') {

        }
    }

    /**
     * Based on the given instructions type, this method will
     * insert them into their corresponding map, e.g. <dynamics>
     * elements will be inserted to <dynamicsMap>. After inserting
     * the map will sorted by `date`.
     * 
     * @param instruction 
     * @param part a part number or 'global'
     */
    insertInstructions<T>(instructions: AnyInstruction[], part: Part) {
        if (!instructions.length) return

        const instructionType = instructions[0].type
        const correspondingMapName = {
            dynamics: 'dynamicsMap',
            ornament: 'ornamentationMap',
            tempo: 'tempoMap'
        }[instructionType]

        if (!correspondingMapName) return

        let map
        if (part === 'global') {
            map = this.rawMPM.performance.global.dated[correspondingMapName]
        }
        else if (typeof part === 'number') {
            map = this.rawMPM.performance.part.find((p: any) => p['@'].number === part+1)
        }

        if (!map) {
            console.log('cannot find part', part, 'in the MPM')
            return
        }

        if (Array.isArray(map[instructionType])) {
            map[instructionType] = [...map[instructionType], ...instructions.map(i => ({'@': i}))]
        }
        else {
            map[instructionType] = instructions.map(i => ({'@': i}))
        }
    }

    /**
     * Will remove the contents of a given map type.
     */
    removeInstructions<T extends DatedInstruction<string>>(part: Part) {
    }

    /**
     * Gets all instructions inside a given map type.
     * @param part 
     * @returns 
     */
    getInstructions<T extends DatedInstruction<string>>(part: Part): T[] {
        return []
    }

    setPerformanceName(performanceName: string) {
        this.rawMPM.performance["@"].name = performanceName
    }

    serialize() {
        return parse('mpm', this.rawMPM)
    }
}
