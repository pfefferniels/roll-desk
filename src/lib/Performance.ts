import { PianoRoll, pitchToSitch } from 'alignmenttool';
import { AnyEvent, MidiFile, NoteOffEvent, NoteOnEvent, SetTempoEvent } from 'midifile-ts';
import { Visitable } from './visitors/Visitable';
import { Visitor } from './visitors/Visitor';

export type MidiNote = {
    id: number,
    onsetTime: number,
    pitch: number,
    velocity: number,
    duration: number
}

/**
 * wrapper for a raw MIDI performance
 */
export class RawPerformance implements Visitable {
    midi: MidiFile | null

    constructor(performance: MidiFile) {
        this.midi = performance
    }

    /**
     * Returns the first and the last onset time of this performance.
     * @returns pair of numbers with lower onset and upper onset
     */
    public range(): [number, number] {
        const performedNotes = this.asNotes()
        const firstOnset = performedNotes[0].onsetTime
        const lastOnset = performedNotes[performedNotes.length - 1].onsetTime
        return [firstOnset, lastOnset]
    }

    private getTickTimeForTempoEvent(event: SetTempoEvent) {
        const microsecondsPerBeat = event.microsecondsPerBeat
        if (!microsecondsPerBeat || !this.midi) return -1

        const BPM = 60000000 / microsecondsPerBeat
        const realTimePerBeat = 60 / BPM
        const ticksPerBeat = this.midi.header.ticksPerBeat
        return realTimePerBeat / ticksPerBeat
    }

    public asNotes(): MidiNote[] {
        if (!this.midi) return []

        const isNoteOn = (event: AnyEvent) => (event as NoteOnEvent).subtype === "noteOn"
        const isNoteOff = (event: AnyEvent) => (event as NoteOffEvent).subtype === "noteOff"
        const isSetTempo = (event: AnyEvent) => (event as SetTempoEvent).subtype === "setTempo"

        let result: MidiNote[] = []
        let currentTime = 0
        let currentTickTime = 0

        this.midi.tracks.forEach((events: AnyEvent[]) => {
            events.forEach((event: AnyEvent, index: number) => {
                currentTime += event.deltaTime

                if (isSetTempo(event)) {
                    currentTickTime = this.getTickTimeForTempoEvent(event as SetTempoEvent)
                }
                else if (isNoteOn(event)) {
                    const noteOnEvent = event as NoteOnEvent

                    result.push({
                        id: index,
                        onsetTime: currentTickTime * currentTime,
                        pitch: noteOnEvent.noteNumber || 0,
                        velocity: noteOnEvent.velocity || 0,
                        duration: 0
                    })
                }
                else if (isNoteOff(event)) {
                    const noteOffEvent = event as NoteOffEvent
                    const correspNote = result.slice().reverse().find(note => note.pitch === noteOffEvent.noteNumber)
                    if (!correspNote) {
                        console.log('no corresponding noteOn event found for noteOff', noteOffEvent)
                        return
                    }
                    correspNote.duration = (currentTickTime * currentTime) - correspNote.onsetTime
                }
            })
        })

        return result
    }

    public asPianoRoll(): PianoRoll {
        const pr = new PianoRoll()
        pr.events = this.asNotes().map(note => {
            return {
                ontime: note.onsetTime,
                offtime: note.onsetTime + note.duration,
                id: note.id.toString(),
                pitch: note.pitch,
                sitch: pitchToSitch(note.pitch),
                onvel: 80,
                offvel: 80,
                channel: 1,
                endtime: note.onsetTime + note.duration,
                label: note.id.toString()
            }
        })

        return pr
    }

    public getById(id: number): MidiNote | undefined {
        return this.asNotes().find(note => note.id === id)
    }

    public accept(visitor: Visitor) {
        return visitor.visitPerformance(this)
    }
}

