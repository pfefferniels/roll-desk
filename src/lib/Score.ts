import { vrvToolkit } from "../components/Verovio"

export type Note = {
    index: number,
    id: string,
    qstamp: number,
    pitch: number,
    duration: number,
    part: number
}

export class Score {
    notes: Note[]
    timemap: any[]  // TODO define type. 

    // score encoding can be anything that Verovio can parse
    constructor(scoreEncoding?: string) {
        console.log('new score object created using Verovio version', vrvToolkit.getVersion(), scoreEncoding)
        vrvToolkit.loadData(scoreEncoding || '')
        this.timemap = vrvToolkit.renderToTimemap()
        this.notes = Score.parseFromTimemap(this.timemap)
    }

    // transform timemap to notes array
    private static parseFromTimemap(timemap: any[]): Note[] {
        let result: Note[] = []
        let index = 0
        for (const event of timemap) {
            if (!event.on) continue
            for (const on of event.on) {
                const midiValues = vrvToolkit.getMIDIValuesForElement(on)
                const offTime = timemap.find((event: any) => event.off && event.off.includes(on)).qstamp || 0
                result.push({
                    index: index, 
                    id: on,
                    qstamp: event.qstamp, 
                    pitch: midiValues.pitch,
                    duration: offTime - event.qstamp,
                    part: 0      // TODO
                })
                index += 1
            }
        }
        console.log('notes array:', result)
        return result
    }

    public asSVG(): string {
        vrvToolkit.setOptions({
            adjustPageHeight: true
        })
        return vrvToolkit.renderToSVG(1)
    }

    public qstampToTstamp(qstamp: number): number {
        return qstamp * 720
    }

    // get last qstamp
    public getMaxQstamp(): number {
        return this.timemap.at(-1).qstamp;
    }

    public notesInRange(start: number, end: number): Note[] {
        const lowerIndex = this.notes.findIndex((note: Note) => note.qstamp >= start)
        const upperIndex = this.notes.findIndex((note: Note) => note.qstamp <= end)
        if (lowerIndex > upperIndex) return []
        return this.notes.slice(lowerIndex, upperIndex)
    }

    public allNotes(): Note[] {
        return this.notes
    }

    public allDownbeats(): number[] {
        return []
    }

    public notesAtTime(qstamp: number): Note[] {
        return this.notes.filter((note: Note) => note.qstamp === qstamp)
    }

    public at(id: string): Note | undefined {
        //return this.notes.find(note => note.index === index)!
        return this.notes.find((value: Note) => value.id === id)
    }
}
