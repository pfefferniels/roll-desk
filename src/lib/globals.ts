import isNode from "detect-node";
import createVerovioModule from 'verovio/wasm';
import { VerovioToolkit } from 'verovio/esm';

export function uuid() {
    return 'axxxxx'.replace(/[x]/g, function (c) {
        var r = Math.random() * 16 | 0, v = c === 'x' ? r : (r & 0x3 | 0x8);
        return v.toString(16);
    });
}

export async function loadDomParser(): Promise<DOMParser> {
    if (isNode) {
        const jsdom = await import('jsdom')
        const JSDOM = new jsdom.JSDOM(``)
        return new JSDOM.window.DOMParser()
    }
    else {
        return new window.DOMParser()
    }
}

export async function loadVerovio(): Promise<any> {
    if (isNode) {
        const module = await createVerovioModule()
        return new VerovioToolkit(module)
    }
    else {
        return window.vrvToolkit
    }
}
