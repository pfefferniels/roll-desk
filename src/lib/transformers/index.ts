export * from "./InterpolateAsynchrony"
export * from "./InterpolateArticulation"
export * from "./InterpolateDynamicsMap"
export * from "./InterpolatePhysicalOrnamentation"
export * from "./InterpolateSymbolicOrnamentation"
export * from "./InterpolateTempoMap"
export * from "./InterpolateTimingImprecision"
export * from "./defaultPipelines"
export * from "./Pipeline"
