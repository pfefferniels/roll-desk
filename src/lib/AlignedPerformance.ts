import { MidiNote, RawPerformance } from "./Performance"
import { Note, Score } from "./Score"
import { SeqNode, Aligner, AlignmentPair, AlignType, GenericSeqNode } from "sequence-align"

type Range = [number, number]
function convertRange(value: number, r1: Range, r2: Range) {
    return (value - r1[0]) * (r2[1] - r2[0]) / (r1[1] - r1[0]) + r2[0];
}

export enum Motivation {
    ExactMatch = "ExactMatch",
    Addition = "Additon",
    Ornamentation = "Ornamentation",
    OctaveAlteration = "OctaveAlteration",
    OctaveDoubling = "OctaveDoubling",
    Error = "Error",
    Omission = "Omission",
    Uncertain = "Uncertain"
}

export type SemanticAlignmentPair = {
    scoreNote?: Note,
    midiNote?: MidiNote,
    motivation: Motivation
}

/**
 * Note in a sequence
 * 
 * @member label id of the note
 * @member featureVector feature vector: [pitch, onsetTime]
 */
class NoteNode extends GenericSeqNode<string> {
    featureVector: number[]

    constructor(label: string, featureVector: number[], transTime = 0) {
        super(label, transTime)
        this.featureVector = featureVector
    }

    /**
     * compute similarity between two note vector representations
     * @param other the other note to compare with
     * @returns similarity score
     */
    computeSimilarity(other: NoteNode): number {
        function gaussDensity(x: number, m: number, sigma: number) {
            return (1 / (Math.sqrt(2 * Math.PI * sigma ** 2))) * Math.exp(-(((x - m) ** 2) / (2 * sigma ** 2)))
        }

        const pitchDistance = Math.abs(this.featureVector[0] - other.featureVector[0])
        const pitchScore = 1 / (pitchDistance + 1)

        const timeScore = convertRange(
            gaussDensity(other.featureVector[1], this.featureVector[1], 3),
            [0, 1],
            [-1, 1])

        return timeScore
    }
}

export class AlignedPerformance {
    score?: Score
    rawPerformance?: RawPerformance
    aligner: Aligner<string>
    rawPairs: AlignmentPair<string>[]
    semanticPairs: SemanticAlignmentPair[]
    gapOpen: number
    gapExt: number

    private byPitch(a: NoteNode, b: NoteNode): number {
        return (a.featureVector[0] - b.featureVector[0]) || (b.featureVector[1] - a.featureVector[1])
    }

    private generateSeqNodesFromScore(notes: Note[], fitIntoRange: [number, number]): NoteNode[] {
        if (notes.length === 0) return []

        const firstQstamp = notes[0].qstamp
        const lastQstamp = notes[notes.length - 1].qstamp
        const scoreRange: Range = [firstQstamp, lastQstamp]

        return notes.map(n => new NoteNode(n.id, [n.pitch, convertRange(n.qstamp, scoreRange, fitIntoRange)]))
    }

    private generateSeqNodesFromPerformance(midiNotes: MidiNote[]): NoteNode[] {
        if (midiNotes.length === 0) return []

        return midiNotes.map((value: MidiNote) => {
            return new NoteNode(value.id.toString(), [value.pitch, value.onsetTime])
        })
    }

    constructor(
        gapOpen?: number,
        gapExt?: number,
        score?: Score,
        rawPerformance?: RawPerformance) {
        this.gapOpen = gapOpen || -1
        this.gapExt = gapExt || -1
        this.score = score
        this.rawPerformance = rawPerformance
        this.aligner = new Aligner<string>(AlignType.Global) // really global
        this.rawPairs = []
        this.semanticPairs = []

        this.performAlignment()
    }

    private performAlignment() {
        if (!this.score || !this.rawPerformance) return

        const performanceNotes = this.rawPerformance.asNotes()
        const scoreNotes = this.score.allNotes()

        if (performanceNotes.length === 0 || scoreNotes.length === 0) return

        // experimental: generate lookup function
        // this might be useful for a more precise time alignment later on.
        {
            const perfNodes = this.generateSeqNodesFromPerformance(performanceNotes)
            const scoreNodes = this.
                generateSeqNodesFromScore(
                    scoreNotes,
                    [performanceNotes[0].onsetTime, performanceNotes[performanceNotes.length - 1].onsetTime])

            let avg1: SeqNode<string>[] = []
            const windowSize = 5
            for (let i = 0; i < perfNodes.length - windowSize; i++) {
                let avg = 0;
                for (let j = 0; j < windowSize; j++) {
                    avg += perfNodes[i + windowSize].featureVector[0]
                }
                avg = Math.pow(windowSize, 1 / avg)
                avg1.push(new SeqNode<string>(perfNodes[i].featureVector[1].toString(), [avg]))
            }

            let avg2: SeqNode<string>[] = []
            for (let i = 0; i < scoreNodes.length - windowSize; i++) {
                let avg = 0;
                for (let j = 0; j < windowSize; j++) {
                    avg += scoreNodes[i + windowSize].featureVector[0]
                }
                avg = Math.pow(windowSize, 1 / avg)
                avg2.push(new SeqNode<string>(scoreNodes[i].featureVector[1].toString(), [avg]))
            }

            this.aligner.align(avg1, avg2, { gapOpen: -0.5, gapExt: -1 })
            const allPairs = this.aligner.retrieveAlignments().paths[0]
        }

        const perfNodes = this.generateSeqNodesFromPerformance(performanceNotes).sort(this.byPitch)
        const scoreNodes = this.
            generateSeqNodesFromScore(
                scoreNotes,
                [performanceNotes[0].onsetTime, performanceNotes[performanceNotes.length - 1].onsetTime]) // scale to range
            .sort(this.byPitch)

        this.aligner.align(perfNodes, scoreNodes, { gapOpen: -1, gapExt: -1 })
        const allPairs = this.aligner.retrieveAlignments().paths[0]
        this.rawPairs.push(...allPairs)

        this.semanticPairs = allPairs.map((pair): SemanticAlignmentPair => {
            const scoreId = pair[1]
            const midiId = pair[0]

            if (scoreId === '-') {
                return {
                    midiNote: this.rawPerformance?.at(+midiId),
                    motivation: Motivation.Addition,
                }
            }
            else if (midiId === '-') {
                return {
                    scoreNote: this.score?.at(scoreId),
                    motivation: Motivation.Omission,
                }
            }
            else {
                return {
                    scoreNote: this.score?.at(scoreId),
                    midiNote: this.rawPerformance?.at(+midiId),
                    motivation: Motivation.ExactMatch
                }
            }
        })
    }

    public setScore(score: Score) {
        this.score = score
        this.performAlignment()
    }

    public setPerformance(performance: RawPerformance) {
        this.rawPerformance = performance
        this.performAlignment()
    }

    public setGapOpen(gapOpen: number) {
        this.gapOpen = gapOpen
    }

    public setGapExt(gapExt: number) {
        this.gapExt = gapExt
    }

    public getAllPairs(): AlignmentPair<string>[] {
        return this.rawPairs
    }

    public setPairs(pairs: AlignmentPair<string>[]) {
        this.rawPairs = pairs
    }

    public getSemanticPairs(): SemanticAlignmentPair[] {
        return this.semanticPairs
    }

    /**
     * This function generates RDF triples out of the Alignment
     * data. 
     * 
     * @todo should it also include the exports of Score and RawPerformance?
     * @returns string of RDF triples in JSON-LD format.
     */
    public serializeToRDF(carriedOutBy: string = ''): string {
        if (!this.semanticPairs.length) return ''

        const result = {
            "@context": {
                // define namespaces here
                "la": "http://example.org/linked_alignment#",
                "crm": "http://www.cidoc-crm.org/cidoc-crm/",
                "xsd": "http://www.w3.org/2001/XMLSchema#",
            },
            "@graph": [
                {
                    "@id": "http://example.org/my-alignment",
                    "@type": "la:Alignment",
                    "crm:P14_carried_out_by": {
                        "@id": `http://example_org/${carriedOutBy}`,
                        "@type": "E39_Actor"
                    },
                    "dcterms:created": new Date(Date.now()).toISOString(),
                    "la:hasAlignmentPair": this.semanticPairs.map((pair, i) => `http//example.org/my-alignment/pair_${i}`)
                },
                ...this.semanticPairs.map((pair, i) => {
                    const result: any = {
                        "@id": `http://example.org/my-alignment/pair_${i}`,
                        "@type": "la:AlignmentPair",
                        "la:hasMotivation": `http://example.org/alignment-motivation#${pair.motivation}`
                    }
                    if (pair.scoreNote) {
                        result["la:hasScoreNote"] = `http://example.org/my-alignment/score.mei#${pair.scoreNote.id}`
                    }
                    if (pair.midiNote) {
                        result["la:hasMIDINote"] = `http://example.org/midi/a-performance/track_0/event_${pair.midiNote.id}`
                    }

                    return result
                }),
            ]
        }

        return JSON.stringify(result, null, 4)
    }

    public noteAtId(id: string) {
        return this.score?.at(id)
    }

    public performedNoteAtIndex(index: number): MidiNote | null {
        return this.rawPerformance?.at(index) || null
    }

    public performedNoteAtId(id: string, part?: number): MidiNote | null {
        const pair = this.rawPairs.find((value: AlignmentPair<string>) => {
            return value[1] === id
        })
        if (pair) return this.rawPerformance?.at(Number(pair[0])) || null
        return null
    }

    public performedNotesAtQstamp(qstamp: number, part?: number): MidiNote[] {
        const notes = this.score?.notesAtTime(qstamp)
        if (!notes) return []

        const midiNotes: MidiNote[] = []
        for (const note of notes) {
            // find this note id in the aligned pairs
            const pair = this.rawPairs.find((pair: AlignmentPair<string>) => pair[1] === note.id)
            if (pair && pair[0] !== '-') {
                const midiNote = this.rawPerformance?.at(Number(pair[0]))
                if (midiNote) midiNotes.push(midiNote)
            }
        }
        return midiNotes
    }

    public ready(): boolean {
        return !!this.score && !!this.rawPerformance
    }

    public qstampOfOnset(onset: number): number {
        if (!this.ready()) return -1

        const noteAtOnset = this.rawPerformance!.asNotes().find(note => note.onsetTime === onset)
        if (!noteAtOnset) return -1

        const matchingPair = this.getAllPairs().find((pair: AlignmentPair<string>) => {
            return pair[0] === noteAtOnset.id.toString()
        })
        if (!matchingPair) return -1

        const note = this.score!.at(matchingPair[1])
        if (!note) return -1

        return note.qstamp
    }

    public tstampOfOnset(onset: number): number {
        return Score.qstampToTstamp(this.qstampOfOnset(onset))
    }
}
