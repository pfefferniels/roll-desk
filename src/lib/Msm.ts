import { Score } from "./Score";
import { AlignedPerformance } from "./AlignedPerformance";
import { Part } from "./Mpm";

/**
 * Temporary attributes used and manipulated in the process of interpolation.
 */
type TemporaryAttributes = {
    readonly 'midi.pitch': number,
    'midi.onset': number,
    'midi.duration': number,
    'midi.velocity': number,
    'bpm'?: number,
    'bpm.beatLength'?: number
}

/**
 * Represents a score note as part of an MSM encoding. 
 * During the process of MPM generation several temporary 
 * attributes will be attached to it.
 */
export type MsmNote = {
    readonly 'xml:id': string,
    readonly 'part': number,
    readonly 'date': number,
    readonly 'duration': number
    readonly pitchname: string
    readonly accidentals: number
    readonly octave: number
} & TemporaryAttributes

/**
 * Used to represent a homophonized version of the score.
 */
export type Chords = {
    [tstamp: number]: MsmNote[]
}

/**
 * This class represents an MSM encoding.
 */
export class MSM {
    allNotes: MsmNote[]

    /**
     * Constructs an MSM representation from a done
     * score-to-performance alignment. 
     * 
     * @param alignedPerformance 
     */
    constructor(alignedPerformance: AlignedPerformance) {
        this.allNotes = alignedPerformance.getSemanticPairs()
            .filter(pair => pair.midiNote && pair.scoreNote)
            .map(pair => {
                return {
                    part: pair.scoreNote!.part,
                    'xml:id': pair.scoreNote!.id,
                    'date': Score.qstampToTstamp(pair.scoreNote!.qstamp),
                    'pitchname': pair.scoreNote!.pname!,
                    'octave': pair.scoreNote!.octave!,
                    'accidentals': pair.scoreNote!.accid!,
                    'midi.pitch': pair.midiNote!.pitch,
                    'midi.onset': pair.midiNote!.onsetTime,
                    'midi.duration': pair.midiNote!.duration,
                    'midi.velocity': pair.midiNote!.velocity,
                    duration: pair.scoreNote!.duration
                }
            })
    }

    public serialize() {
        return {
            'msm': {
                '@': {
                    title: 'aligned performance',
                    pulsesPerQuarter: 720,
                },
                'global': {
                    'header': {},
                    'dated': {
                        'timeSignatureMap': {
                            'timeSignature': {
                                '@': {
                                    'date': 0.0,
                                    'numerator': 3,
                                    'denominator': 4,
                                }
                            }
                        },
                        'sectionMap': {
                            // TODO: derive from FormalAlterations
                            'section': {
                                '@': {
                                    date: 0.0,
                                    'date.end': this.allNotes[this.allNotes.length - 1].date
                                }
                            }
                        },
                    }
                },
                'part': Array.from(Array(2).keys()).map(part => {
                    return {
                        '@': {
                            name: `part${part}`,
                            number: `${part + 1}`,
                            'midi.port.channel': part,
                            'midi.port': 0
                        },
                        header: {},
                        dated: {
                            'keySignatureMap': {
                                'keySignature': {
                                    '@': {
                                        date: 0
                                    }
                                }
                            }
                        },
                        score: {
                            'note': this.allNotes.filter(note => note.part === part)
                        }
                    }
                })
            }
        }
    }

    /**
     * Returns all notes present at a given score date in a given
     * part.
     * @param tstamp score date
     * @param part if "global", all parts will be considered
     * @returns array of MSM notes
     */
    public notesAtDate(tstamp: number, part: Part): MsmNote[] {
        return this.allNotes.filter(note => {
            return (typeof part === 'number') ?
                (note.date === tstamp && note.part === part) // a specific part
                : (note.date === tstamp) // consider all parts
        })
    }

    /**
     * Generates a homophonized version of the MSM score.
     * @returns 
     */
    public asChords(): Chords {
        return this.allNotes.reduce((prev: any, curr) => {
            if (prev[curr.date]) {
                prev[curr.date].push(curr)
            }
            else {
                prev[curr.date] = [curr]
            }
            return prev
        }, {})
    }

    /**
     * Returns the last date, at which a note is present.
     * @returns score date in ticks
     */
    public lastDate(): number {
        return Math.max(...this.allNotes.map(note => note.date))
    }
}
