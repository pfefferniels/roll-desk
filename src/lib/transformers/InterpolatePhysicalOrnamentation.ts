import { MPM, Ornament } from "../Mpm"
import { MSM } from "../Msm"
import { AbstractTransformer, TransformationOptions } from "./Transformer"

export interface InterpolatePhysicalOrnamentationOptions extends TransformationOptions {
    /**
     * the minimum number of notes an arpeggio is expected to have (inclusive)
     */
    minimumArpeggioSize: number
}

/**
 * Interpolates arpeggiated chords as ornaments, inserts them as physical
 * values into the MPM and substracts accordingly from the MIDI onset, so
 * that after the transformation all notes of the chord will have the same
 * onset.
 */
export class InterpolatePhysicalOrnamentation extends AbstractTransformer<InterpolatePhysicalOrnamentationOptions> {
    public name() { return 'InterpolatePhysicalOrnamentation' }

    public transform(msm: MSM, mpm: MPM): string {
        const isSorted = (arr: number[]) => {
            let direction = -(arr[0] - arr[1])
            for (let [i, val] of arr.entries()) {
                direction = !direction ? -(arr[i - 1] - arr[i]) : direction
                if (i === arr.length - 1)
                    return !direction ? 0 : direction / Math.abs(direction)
                else if ((val - arr[i + 1]) * direction > 0) return 0
            }
        }

        const ornaments: Ornament[] = []

        const chords = msm.asChords()
        for (const [date, arpeggioNotes] of Object.entries(chords)) {
            if (arpeggioNotes.length >= (this.options?.minimumArpeggioSize || 2)) {
                const sortedByOnset = arpeggioNotes.sort((a, b) => a['midi.onset'] - b['midi.onset'])

                const arpeggioDirection = isSorted(sortedByOnset.map(note => note["midi.pitch"]))
                let noteOrder = ''
                if (arpeggioDirection === 1) noteOrder = 'ascending pitch'
                else if (arpeggioDirection === -1) noteOrder = 'descending pitch'
                else noteOrder = sortedByOnset.map(note => `#${note["xml:id"]}`).join(' ')
                
                const duration = sortedByOnset[sortedByOnset.length-1]["midi.onset"] - sortedByOnset[0]["midi.onset"] 

                ornaments.push({
                    'type': 'ornament',
                    'date': +date,
                    'name.ref': 'neutralArpeggio',
                    'note.order': noteOrder,
                    'frame.start': (-duration/2) * 1000,
                    'frameLength': duration * 1000,
                    'scale': 0.0,
                    'time.unit': 'milliseconds'
                })

                const onsetSum = arpeggioNotes.map(note => note['midi.onset']).reduce((a, b) => a + b, 0)
                const avarageOnset = (onsetSum / arpeggioNotes.length) || 0

                arpeggioNotes.forEach(note => {
                    note['midi.onset'] = avarageOnset
                })
            }
        }

        mpm.insertInstructions(ornaments, 'global')

        // hand it over to the next transformer
        return super.transform(msm, mpm)
    }
}
