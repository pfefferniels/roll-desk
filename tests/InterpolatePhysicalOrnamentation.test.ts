import { MPM, Ornament } from "../src/lib/Mpm"
import { MSM } from "../src/lib/Msm"
import { AlignedPerformance } from "../src/lib/AlignedPerformance"
import { Mei } from "../src/lib/Score"
import { RawPerformance } from "../src/lib/Performance"
import { InterpolatePhysicalOrnamentation } from "../src/lib/transformers/InterpolatePhysicalOrnamentation"
import * as fs from 'fs';
import { loadVerovio, loadDomParser } from '../src/lib/globals'

const generateMSM = async (meiFile: string, midiFile: string): Promise<MSM> => {
    let { read } = await import("midifile-ts");

    const mei = fs.readFileSync(meiFile, 'utf16le')
    const midi = fs.readFileSync(midiFile)
    const arr = Uint8Array.from(midi)

    const score = new Mei(mei, await loadVerovio(), await loadDomParser())
    const performance = new RawPerformance(read(arr))
    const alignedPerformance = new AlignedPerformance(score, performance)
    return new MSM(alignedPerformance)
}

describe('InterpolatePhysicalOrnamentation', () => {
    it(`does not interpolate anything when no ornamentation is given`, async () => {
        const msm = await generateMSM('tests/files/test000/score.mei', 'tests/files/test000/neutral.mid')
        const mpm = new MPM()

        const transformer = new InterpolatePhysicalOrnamentation()
        transformer.transform(msm, mpm)

        const ornamentInstruction = mpm.getInstructions<Ornament>('ornament', 'global')
        console.log('ornamentInstructions=', ornamentInstruction)

        //expect(ornamentInstruction.length).toEqual(0)
    })
})
