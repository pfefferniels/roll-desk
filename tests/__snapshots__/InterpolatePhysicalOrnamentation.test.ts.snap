// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InterpolatePhysicalOrnamentation correctly interpolates complex arpeggiations (WM 79) 1`] = `
Object {
  "date": 0,
  "frame.start": -65.10416666666674,
  "frameLength": 130.20833333333348,
  "gradient": "crescendo",
  "name.ref": "neutralArpeggio",
  "note.order": "#m-58 #m-59 #m-60 #m-48 #m-47 #m-44",
  "scale": 2.5,
  "time.unit": "milliseconds",
  "type": "ornament",
  "xml:id": Any<String>,
}
`;

exports[`InterpolatePhysicalOrnamentation correctly interpolates complex arpeggiations (WM 79) 2`] = `
Object {
  "date": 720,
  "frame.start": -17.708333333333435,
  "frameLength": 35.41666666666687,
  "gradient": "no-gradient",
  "name.ref": "neutralArpeggio",
  "note.order": "#m-62 #m-51 #m-63 #m-50 #m-64",
  "scale": 0,
  "time.unit": "milliseconds",
  "type": "ornament",
  "xml:id": Any<String>,
}
`;

exports[`InterpolatePhysicalOrnamentation correctly interpolates complex arpeggiations (WM 79) 3`] = `
Object {
  "date": 1440,
  "frame.start": -49.479166666666515,
  "frameLength": 98.95833333333303,
  "gradient": "crescendo",
  "name.ref": "neutralArpeggio",
  "note.order": "#m-66 #m-68 #m-67 #m-53 #m-54",
  "scale": 0.5,
  "time.unit": "milliseconds",
  "type": "ornament",
  "xml:id": Any<String>,
}
`;

exports[`InterpolatePhysicalOrnamentation correctly interpolates complex arpeggiations (WM 79) 4`] = `
Object {
  "date": 2160,
  "frame.start": -155.2083333333334,
  "frameLength": 310.4166666666668,
  "gradient": "crescendo",
  "name.ref": "neutralArpeggio",
  "note.order": "ascending pitch",
  "scale": 7.5,
  "time.unit": "milliseconds",
  "type": "ornament",
  "xml:id": Any<String>,
}
`;

exports[`InterpolatePhysicalOrnamentation correctly interpolates complex arpeggiations (WM 79) 5`] = `
Object {
  "date": 2880,
  "frame.start": -68.22916666666634,
  "frameLength": 136.4583333333327,
  "gradient": "crescendo",
  "name.ref": "neutralArpeggio",
  "note.order": "#m-94 #m-93 #m-95 #m-80 #m-81 #m-77",
  "scale": 4.5,
  "time.unit": "milliseconds",
  "type": "ornament",
  "xml:id": Any<String>,
}
`;

exports[`InterpolatePhysicalOrnamentation correctly interpolates complex arpeggiations (WM 79) 6`] = `
Object {
  "date": 3600,
  "frame.start": -93.75,
  "frameLength": 187.5,
  "gradient": "crescendo",
  "name.ref": "neutralArpeggio",
  "note.order": "#m-97 #m-98 #m-84 #m-99 #m-82",
  "scale": 3,
  "time.unit": "milliseconds",
  "type": "ornament",
  "xml:id": Any<String>,
}
`;

exports[`InterpolatePhysicalOrnamentation correctly interpolates complex arpeggiations (WM 79) 7`] = `
Object {
  "date": 4320,
  "frame.start": -85.9375,
  "frameLength": 171.875,
  "gradient": "crescendo",
  "name.ref": "neutralArpeggio",
  "note.order": "ascending pitch",
  "scale": 2.5,
  "time.unit": "milliseconds",
  "type": "ornament",
  "xml:id": Any<String>,
}
`;
